feanil/reorg guides

feanil opened the pull request:
Note: Review https://github.com/openedx/edx-platform/pull/32611 and https://github.com/openedx/edx-platform/pull/32612 before this one.

These docs aren't currently being published anywhere, so you'll have to generate them locally to test.

You can generate them by running the following in an environment where the LMS can come up:

```
cd docs/guides
make html
open _build/html/index.html
```


- docs: Add the new sphinx-design extension.
- chore: Run `make compile-requirements` to add sphinx-design.
- docs: Move how-tos and references to sub-folders.
- docs: Rename guides to concepts.
- docs:  Make the generated docs visible.

feanil committed:
docs: Add the new sphinx-design extension.

This gives us the abuliity to add some nice responsive web components to
our docs. In particular girds and cards.

feanil committed:
chore: Run `make compile-requirements` to add sphinx-design.

feanil committed:
docs: Move how-tos and references to sub-folders.

feanil committed:
docs: Rename guides to concepts.

The guides map pretty closely to the concept docs in diataxis so rename
them as such to align to the standard naming we're moving to.

feanil committed:
docs:  Make the generated docs visible.

There is still a lot of improvements that could be made to make the
generated docs better organized.  However, before this change we were
generating the docs but they weren't in the doc tree so you couldn't
actually browse to them.  So we fix that first so that we can start
getting feedback about how to organize them.

feanil commented:
@openedx/arch-bom @kdmccormick this one is ready for review.

kdmccormick commented:
I'll take a look this afternoon.

kdmccormick commented on requirements/edx/doc.in:
```diff
@@ -5,3 +5,4 @@ code-annotations          # provides annotations for certain documentation
 sphinx-book-theme         # Common theme for all Open edX projects
 gitpython                 # fetch git repo information
 Sphinx                    # Documentation builder
+sphinx-design             # provides various responsive web-components
```
```suggestion
sphinx-design             # provides various responsive web-components
sphinxcontrib-openapi[markdown]
```
Using base requirements, in order to get `make html` to pass, I also had to install `sphinxcontrib-openapi[markdown]`, so you might want to add that package here. 

Although, for what it's worth, that package does already exist in development.in. Not sure whether you see doc.in as parallel to development.in, or an extension of development.in.

feanil replied:
Added below with a comment.

kdmccormick commented on docs/guides/index.rst:
```diff
@@ -24,17 +24,64 @@ locations.
 Change History
 **************
 
+* Jun 30, 2023
```
```suggestion
* Jun 30, 2023

```
so that the nested list renders

kdmccormick reviewed (commented):
The references/docstrings compilation is taking a while. 20 minutes and it's not even halfway through o_O. Do you recall it being that slow? Could be that fact that I'm trying to go through Tutor.

A couple comments for you in the meantime.

kdmccormick reviewed (approved):
Looks good, pending my two suggestions above.

Probably out of scope of this PR, but just to throw the idea out there, maybe we should axe the docstrings reference. It feels like a worse version of just browsing the code, and it took ~40mins to build for me (albeit, through a bind-mounted Docker volume).

robrap commented:
> Probably out of scope of this PR, but just to throw the idea out there, maybe we should axe the docstrings reference. It feels like a worse version of just browsing the code, and it took ~40mins to build for me (albeit, through a bind-mounted Docker volume).

Not saying we should or we shouldn't, but I've always wondered where to point when wanting to provide a link to a class, because in github, I can either provide a permanent link (even thought the class may move in the file), or point to the current version of the file, and force someone to search for the name in the file. Alternatively, someone could link to the class in the docs, which should just work as long as the class didn't move. No solution is ideal. Even a github search might be new vs old search and work for some and not for all. :)

feanil committed:
build: Re-org docs deps.

Some doc buildig deps were in development.in explicitly while others
weren't.  Update development.in to pull in doc.txt and then remove
direct additions of the doc requirements from development.in

kdmccormick commented:
New docs deps structure looks good 👍🏻

robrap commented:
Non-blocking, but wondering if you could post before/after images of the doc outline? Thanks.

feanil commented:
Before: 
![guides-before](https://github.com/openedx/edx-platform/assets/781561/34d2edee-4551-4db1-988b-d84a189d1d91)

After:
![guides-after](https://github.com/openedx/edx-platform/assets/781561/1bb1a55a-e8a3-44a9-bc49-83502542b921)

robrap commented:
I'm excited to see this work finally being done. Thanks so much @feanil.
